torch/optim/optimizer.py

```diff
@@ -135,9 +135,9 @@ def cast(param, value):
         for k, v in state_dict['state'].items():
             if k in id_map:
                 param = id_map[k]
-                state[param] = cast(param, v)
+                state[id(param)] = cast(param, v)
```

Comment:
We might need to fix `__setstate__` in case someone saved optimizers from before this patch.

Reply:
If the only way to save optimizers is via `state_dict`, then I don't think there is an issue. Keys for `state_dict` are `id`s, and hence there should not be an issue.

Am I missing something?

Reply:
@apaszke I have just confirmed that this is rather not an issue. I saved the `state_dict`  of an `Adam` instance prior to this patch, and was able to load it successfully with the changes made in this patch and the other way round too.

Reply:
I believe the point @apaszke made was that if someone saved the full `optimizer` object (which some people unfortunately do), then this might break their code.

Reply:
Yes, the problem is with whole objects being serialized.

Reply:
Thank you @fmassa @apaszke . I'll rebase and push in the required changes as soon as I can.